File: app/src/test/java/dev/suhockii/lifetest/repo/RemoteRepositoryTest.java
package dev.suhockii.lifetest.repo;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.Mockito;

import java.util.ArrayList;

import dev.suhockii.lifetest.data.local.dao.ProductDao;
import dev.suhockii.lifetest.data.local.dao.ProductDetailsDao;
import dev.suhockii.lifetest.data.local.entity.ProductEntity;
import dev.suhockii.lifetest.data.remote.ProductsApi;
import dev.suhockii.lifetest.data.remote.entity.ProductDetailsResponse;
import dev.suhockii.lifetest.data.remote.entity.ProductListResponse;
import dev.suhockii.lifetest.model.Product;
import io.reactivex.Observable;
import io.reactivex.Single;

import static org.junit.Assert.*;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(JUnit4.class)
public class RemoteRepositoryTest {
    private ProductsApi productsApi;
    private AppRepository remoteRepository;

    @Before
    public void setUp() throws Exception {
        productsApi = Mockito.mock(ProductsApi.class);
        remoteRepository = new RemoteRepository(productsApi);
    }

    @Test
    public void getProducts() throws Exception {
        when(productsApi.getProducts()).thenReturn(Observable.just(new ProductListResponse()));

        remoteRepository.getProducts();

        verify(productsApi, times(1)).getProducts();
    }

    @Test
    public void getDetailsFor() throws Exception {
        String productId = "1";
        Product product = new ProductEntity(productId);
        Single<ProductDetailsResponse> detailsResponse =
                Single.just(new ProductDetailsResponse(productId));

        when(productsApi.getProductDetails(product.getId())).thenReturn(detailsResponse);

        remoteRepository.getDetailsFor(product);

        verify(productsApi, times(1)).getProductDetails(productId);
    }

}

File: app/src/test/java/dev/suhockii/lifetest/repo/RemoteRepositoryTest.kt
package dev.suhockii.lifetest.repo

import dev.suhockii.lifetest.data.local.entity.ProductEntity
import dev.suhockii.lifetest.data.remote.ProductsApi
import dev.suhockii.lifetest.data.remote.entity.ProductDetailsResponse
import dev.suhockii.lifetest.data.remote.entity.ProductListResponse
import io.reactivex.Observable
import io.reactivex.Single
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.Mockito
import org.mockito.Mockito.*

@RunWith(JUnit4::class)
class RemoteRepositoryTest {
    private var productsApi: ProductsApi? = null
    private var remoteRepository: AppRepository? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        productsApi = Mockito.mock(ProductsApi::class.java)
        remoteRepository = RemoteRepository(productsApi!!)
    }

    @Test
    @Throws(Exception::class)
    fun getProducts() {
        `when`(productsApi!!.products).thenReturn(Observable.just(ProductListResponse()))

        remoteRepository!!.getProducts()

        verify<ProductsApi>(productsApi, times(1)).products
    }

    @Test
    @Throws(Exception::class)
    fun getDetailsFor() {
        val productId = "1"
        val product = ProductEntity(productId)
        val detailsResponse = Single.just(ProductDetailsResponse(productId))

        `when`(productsApi!!.getProductDetails(product.id)).thenReturn(detailsResponse)

        remoteRepository!!.getDetailsFor(product)

        verify<ProductsApi>(productsApi, times(1)).getProductDetails(productId)
    }

}